crypto/hpke/hpke.c

```diff
@@ -0,0 +1,2178 @@
+                                            OSSL_KEM_PARAM_OPERATION_DHKEM,
+                                            0);
+    *p = OSSL_PARAM_construct_end();
+    if (ctx->mode == OSSL_HPKE_MODE_AUTH
+        || ctx->mode == OSSL_HPKE_MODE_PSKAUTH) {
+        int kem_ind = 0;
+
+        kem_ind = kem_iana2index(ctx->suite.kem_id);
+        if (kem_ind == 0) {
+            erv = 0;
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+            goto err;
+        }
+        if (hpke_kem_id_nist_curve(ctx->suite.kem_id) == 1) {
+            spub = EVP_PKEY_new_raw_nist_public_key(ctx->libctx, ctx->propq,
+                                                    hpke_kem_tab[kem_ind].
+                                                    groupid,
+                                                    hpke_kem_tab[kem_ind].
+                                                    groupname,
+                                                    ctx->authpub,
+                                                    ctx->authpublen);
+        } else {
+            spub = EVP_PKEY_new_raw_public_key_ex(ctx->libctx,
+                                                  hpke_kem_tab[kem_ind].keytype,
+                                                  ctx->propq,
+                                                  ctx->authpub,
+                                                  ctx->authpublen);
+        }
+        if (spub == NULL) {
+            erv = 0;
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+            goto err;
+        }
+        if (EVP_PKEY_auth_decapsulate_init(pctx, spub, params) != 1) {
+            erv = 0;
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+            goto err;
+        }
+    } else {
+        if (EVP_PKEY_decapsulate_init(pctx, params) != 1) {
+            erv = 0;
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+            goto err;
+        }
+    }
+    erv = EVP_PKEY_decapsulate(pctx, NULL, &lsslen, enc, enclen);
+    if (erv != 1) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    if (lsslen > OSSL_HPKE_MAXSIZE) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    erv = EVP_PKEY_decapsulate(pctx, lss, &lsslen, enc, enclen);
+    if (erv != 1) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    EVP_PKEY_CTX_free(pctx);
+    pctx = NULL;
+    /* free shared_secret in case this is 2nd call */
+    OPENSSL_clear_free(ctx->shared_secret, ctx->shared_secretlen);
+    ctx->shared_secret = OPENSSL_malloc(lsslen);
+    if (ctx->shared_secret == NULL) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    ctx->shared_secretlen = lsslen;
+    memcpy(ctx->shared_secret, lss, lsslen);
+
+err:
+    EVP_PKEY_CTX_free(pctx);
+    EVP_PKEY_free(spub);
+    OPENSSL_cleanse(lss, sizeof(lss));
```

Comment:
If you set lsslen to zero at init then you can use that values here.. in the cleanse

Reply:
could do that yes, question though: for a fixed size buffer on the stack like ``lss`` here, wouldn't it be (very very slightly:-) better to cleanse the entire thing and not only the octets-used? just thinking about potential timing signals leaking (not that this change would be significant though)